rust/chains/hyperlane-cosmos/src/aggregation_ism.rs

```diff
@@ -0,0 +1,81 @@
+use crate::{
+    grpc::{WasmGrpcProvider, WasmProvider},
+    payloads::aggregate_ism::{
+        ModulesAndThresholdRequest, ModulesAndThresholdRequestInner, ModulesAndThresholdResponse,
+    },
+    verify::bech32_decode,
+    ConnectionConf, CosmosProvider, Signer,
+};
+use async_trait::async_trait;
+use hyperlane_core::{
+    AggregationIsm, ChainResult, ContractLocator, HyperlaneChain, HyperlaneContract,
+    HyperlaneDomain, HyperlaneMessage, HyperlaneProvider, RawHyperlaneMessage, H256,
+};
+use tracing::instrument;
+
+/// A reference to an AggregationIsm contract on some Cosmos chain
+#[derive(Debug)]
+pub struct CosmosAggregationIsm {
+    _conf: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    _signer: Signer,
+    provider: Box<WasmGrpcProvider>,
+}
+
+impl CosmosAggregationIsm {
+    /// create new Cosmos AggregationIsm agent
+    pub fn new(conf: ConnectionConf, locator: ContractLocator, signer: Signer) -> Self {
+        let provider = WasmGrpcProvider::new(conf.clone(), locator.clone(), signer.clone());
+
+        Self {
+            _conf: conf,
+            domain: locator.domain.clone(),
+            address: locator.address,
+            _signer: signer,
+            provider: Box::new(provider),
+        }
+    }
+}
+
+impl HyperlaneContract for CosmosAggregationIsm {
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl HyperlaneChain for CosmosAggregationIsm {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(CosmosProvider::new(self.domain.clone()))
+    }
+}
+
+#[async_trait]
+impl AggregationIsm for CosmosAggregationIsm {
+    #[instrument(err)]
+    async fn modules_and_threshold(
+        &self,
+        message: &HyperlaneMessage,
+    ) -> ChainResult<(Vec<H256>, u8)> {
+        let payload = ModulesAndThresholdRequest {
```

Comment:
wonder if we should just put a constructor on ModulesAndThresholdRequest that takes in a &HyperlaneMessage